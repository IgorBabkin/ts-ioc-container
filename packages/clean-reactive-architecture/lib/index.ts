export class MyNumber {
  constructor(private value: number) {}

  add(value: number): void {
    this.value += value;
  }

  sub(value: number): void {
    this.value -= value;
  }

  getValue(): number {
    return this.value;
  }
}
