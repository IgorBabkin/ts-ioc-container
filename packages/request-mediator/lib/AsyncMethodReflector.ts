export class AsyncMethodReflector {
    constructor(readonly hookKey: string | symbol) {}

    // eslint-disable-next-line @typescript-eslint/ban-types
    private addHook<GInstance extends Object>(target: GInstance, propertyKey: string | symbol): void {
        const targetId = AsyncMethodReflector.getTargetId(target);
        const hooks = Reflect.getMetadata(this.hookKey, targetId) || [];
        Reflect.defineMetadata(this.hookKey, [...hooks, propertyKey], targetId);
    }

    // eslint-disable-next-line @typescript-eslint/ban-types
    async invokeHooksOf<GInstance extends Object>(target: GInstance): Promise<void> {
        const targetId = AsyncMethodReflector.getTargetId(target);
        const hooks: string[] = Reflect.hasMetadata(this.hookKey, targetId)
            ? Reflect.getMetadata(this.hookKey, targetId)
            : [];

        for (const hookMethod of hooks) {
            await (target as any)[hookMethod]();
        }
    }

    // eslint-disable-next-line @typescript-eslint/ban-types
    private static getTargetId<GInstance extends Object = any>(target: GInstance): any {
        return target.constructor;
    }

    createMethodHookDecorator(): MethodDecorator {
        return (target, propertyKey) => {
            // eslint-disable-next-line @typescript-eslint/ban-types
            this.addHook(target, propertyKey);
        };
    }
}
