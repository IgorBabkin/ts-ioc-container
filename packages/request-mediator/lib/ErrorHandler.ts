type constructor<T> = new (...args: any[]) => T;

export interface IErrorHandler<Context = undefined> {
  handle(error: unknown, context: Context): void;
}

export abstract class ErrorHandler<Context> implements IErrorHandler<Context> {
  protected abstract errors: constructor<Error>[];

  constructor(private handler: IErrorHandler<Context>) {}

  handle(error: unknown, context: Context): void {
    return this.matchError(error) ? this.handleError(error, context) : this.handler.handle(error, context);
  }

  protected abstract handleError(error: Error, context: Context): void;

  private matchError(error: unknown): error is Error {
    return this.errors.some((it) => {
      return error instanceof it;
    });
  }
}
