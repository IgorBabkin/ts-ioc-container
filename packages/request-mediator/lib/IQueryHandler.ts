export interface IQueryHandler<TQuery, TResponse> {
    handle(query: TQuery): Promise<TResponse>;
}
