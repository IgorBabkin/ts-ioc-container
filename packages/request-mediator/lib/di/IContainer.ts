import { constructor } from 'ts-constructor-injector';

export interface IContainer {
    registerValue(key: string | symbol, value: unknown): void;

    createScope(tags: string[]): IContainer;

    resolve<T>(QueryHandler: constructor<T> | symbol): T;

    dispose(): void;
}
