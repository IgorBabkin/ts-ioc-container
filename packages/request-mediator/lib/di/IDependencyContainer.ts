import { constructor } from '../others';

export interface IDependencyContainer {
  registerValue(key: string | symbol, value: unknown): void;

  createScope(tags: string[]): IDependencyContainer;

  resolve<T>(key: constructor<T> | symbol): T;

  dispose(): void;

  onBeforeDispose(): Promise<void>;
}
