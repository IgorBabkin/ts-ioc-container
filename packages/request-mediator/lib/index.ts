export { IDependencyContainer } from './di/IDependencyContainer';

export { RequestMediator, request } from './app/RequestMediator';
export { useCase } from './app/UseCaseMediator';
export { useService } from './app/ServiceMediator';
export { Scope } from './app/Scope';
export { ErrorHandler, IErrorHandler } from './ErrorHandler';
export { IQueryHandler } from './IQueryHandler';
export { before, after } from './mediator/SimpleMediator';
export { IMediator } from './mediator/IMediator';
export { ITransactionContext, transaction, ITransactionContextKey } from './mediator/transaction/ITransactionContext';
