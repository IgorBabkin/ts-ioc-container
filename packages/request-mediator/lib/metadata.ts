export const prop =
  <T>(key: string | symbol, value: T): ClassDecorator =>
  (target) => {
    Reflect.defineMetadata(key, value, target);
  };

// eslint-disable-next-line @typescript-eslint/ban-types
export function getProp<T>(target: Object, key: string | symbol): T | undefined {
  return Reflect.getOwnMetadata(key, target) as T;
}
