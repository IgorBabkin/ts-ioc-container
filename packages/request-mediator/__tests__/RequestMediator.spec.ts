import 'reflect-metadata';
import { by, inject, Provider } from 'ts-ioc-container';
import { Context } from './context/Context';
import { after, before, IQueryHandler, request, RequestMediator, useCase, useService } from '../lib';
import { ContainerAdapter, createContainer, EmptyType, onDispose } from './di';

export class Logger extends Context<string[]> {
  addLog(log: string): void {
    const logs = this.getValue();
    this.setValue(logs.concat(log));
  }

  @onDispose
  async save(): Promise<void> {}
}

class QueryHandler2 implements IQueryHandler<EmptyType, void> {
  constructor(@inject(by('Logger')) private logger: Logger) {}

  handle(query: EmptyType): Promise<void> {
    this.logger.addLog('QueryHandler2');
    return Promise.resolve(undefined);
  }
}

class QueryHandler1 implements IQueryHandler<EmptyType, void> {
  constructor(@inject(by('Logger')) private logger: Logger) {}

  @before
  async before(): Promise<void> {
    this.logger.addLog('before');
  }

  handle(query: EmptyType): Promise<void> {
    this.logger.addLog('QueryHandler1');
    return Promise.resolve(undefined);
  }

  @after
  async after(): Promise<void> {
    this.logger.addLog('after');
  }
}

@request('before', [QueryHandler1, QueryHandler2])
@request('after', [QueryHandler2, QueryHandler1])
@useCase('before', [QueryHandler1])
class QueryHandler3 implements IQueryHandler<EmptyType, void> {
  constructor(
    @inject(by('Logger')) private logger: Logger,
    @inject(useService(QueryHandler1)) private queryHandler: QueryHandler1,
  ) {}

  async handle(query: EmptyType): Promise<void> {
    this.logger.addLog('QueryHandler3');
    await this.queryHandler.handle(query);
  }
}

describe('RequestMediator', () => {
  it('should invoke middleware', async () => {
    const container = createContainer().register('Logger', Provider.fromValue(new Logger('logger', [])));

    const mediator = new RequestMediator(new ContainerAdapter(container));

    await mediator.send(QueryHandler3, {});

    expect(container.resolve<Logger>('Logger').getValue()).toMatchSnapshot();
  });
});
