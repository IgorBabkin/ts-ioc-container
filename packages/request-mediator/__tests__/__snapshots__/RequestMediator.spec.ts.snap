// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestMediator should invoke middleware 1`] = `
Array [
  "before",
  "QueryHandler1",
  "after",
  "QueryHandler2",
  "QueryHandler3",
  "before",
  "QueryHandler1",
  "after",
  "QueryHandler2",
  "before",
  "QueryHandler1",
  "after",
]
`;
